### ISSUE TITLE
Query:Cython, npy_math.h and complex.h inclusion

### ISSUE THREAD

**Issue by ilayn** [scipy/scipy#19657]
As part of the Cythonization efforts, I have been hitting the common issue of getting access to complex versions of abs, sqrt, log and so on functions. This is problematic for a few reasons, 

- MSVC in windows is c99 incompatible and we all know the history etc. it is what it is
- npy_math is on the pedestal with discussions such as https://github.com/numpy/numpy/issues/20880 and similar discussions elsewhere hence it is only going to be less useful in the future, in my limited understanding
- Cython injects guard macros after the includes and we have to play with the `CCYTHON_COMPLEX` env var and do some custom bricklaying at the Cython compilation setup (if I am understanding correctly) Here is a better discussion of it https://stackoverflow.com/questions/57837255/defining-dcomplex-externally-in-cython

In our legacy code we have a few avenues of "hacks", namely 

- the types switching back and forth between `_DComplex` , `double complex` and `npy_cdouble` types in unions or casting schemes, 
- The imports from `npy_math` as in https://github.com/scipy/scipy/blob/fa9f13e6906e7d00510d593f7f982db30e4e4f14/scipy/special/_complexstuff.pxd#L12-L24
- and some other redefinition clauses of `_Dcomplex` and so on

None of them seems to be the clean way but I'd like to ask our build residents what way we should pursue to have a sustainable code? 

I'll ping Cython, NumPy folks too, so that we don't stray too outwards @da-woods @lysnikolaou @rgommers @seberg

I would really appreciate a mini code block that should run succesfully on a win or *nix system such as 

```python
%%cython -f
import numpy as np
cimport numpy as cnp

# just to make sure that things real and complex funcs can coexist
from libc.math cimport (abs, sqrt)

<insert complex cimport magic here>


cdef double test_func(double complex z):
    cdef double absz, sqaz
    absz = cabs(z)
    sqaz = sqrt(absz)
    return sqaz
```

If you want to make functions run with `npy_cdouble` and `npy_cabs` that's also fine. I just need some clarification which way we should be going towards. 

We can also disallow MSVC but I feel that it's a bit too much of a restriction.

**Comment by rgommers** [comment]
> MSVC in windows is c99 incompatible

It's just the complex types that are differently named, but using C99 `cabs` in C or Cython should work fine AFAIK (MSVC should be fully C99 compliant modulo the mess with types).

This is always a hairy question; I don't quite have the bandwidth to write the requested code snippet here, hopefully @lysnikolaou will be able to do so.

**Comment by ilayn** [comment]
In the meantime I convinced myself that it won't be possible but if @lysnikolaou has the answer that's great

**Comment by da-woods** [comment]
I haven't played too much with complex in *real* code, so a little unclear on the issues (although I totally believe they exist!)

With recent Cython (3.0+) on Linux I just have to add `from libc.complex cimport cabs` and your code compiles for me as written on Linux with GCC. My impression is that `CYTHON_CCOMPLEX` gets detected mostly correctly, and so the `complex double` types can get passed to suitable C functions. I imagine you don't have to go that far back for that not to be true though.

The alternative Numpy `npy_cdouble` types don't look like they'll automatically work though.

So it looks like the basic problem is: you've got a bunch of types that are very close to equivalent (same information, same memory layout, but different name). And perhaps what would be useful to do is have a way of declaring "this is a complex type" in Cython. Maybe assume that they're all compatible with the basic C or C++ standard library memory layout (which I think is true) and let Cython do the automatic conversions between them.

That's probably something we could support better in some future version of Cython, although doesn't help now.

**Comment by ilayn** [comment]
> With recent Cython (3.0+) on Linux I just have to add from libc.complex cimport cabs and your code compiles for me as written on Linux with GCC. 

Yes on windows it fails depending on clang or mscv pickup by the toolchain. 

> Maybe assume that they're all compatible with the basic C or C++ standard library memory layout (which I think is true) and let Cython do the automatic conversions between them.

This would be indeed a great improvement. Not sure how the internals work but instead of the user is doing the conversions in various codebases, a canonical Cython machinery would be quite useful. Currently I'm basically using a double boot machine and testing in both systems Win and Linux and whichever complains I'm adjusting the code and see what the other starts complaining until both does not complain.

**Comment by ilayn** [comment]
Since we have been largely writing plain C/C++ and we are deciding to bite the bullet with `#ifdef _MSC_VER` guards duplicating the complex branches for C code, I'll close this one.
